### DISCUSSION TITLE
Passing data from upgrade to websocket

### DISCUSSION THREAD

**Discussion by streamich** [uNetworking/uWebSockets.js#365] [Q&A] [answered]
I'm trying to pass headers from the upgrade callback to the WebSocket object, so that I can access those headers on websocket "open" and "message" callbacks.

For example, let's say I want to access a header called `Super-Header` in "message" callback, I'm looking for something like this:

```js
  {
      upgrade: (res, req, context) => {
        context.mySuperHeader = req.getHeader('super-header');

        res.upgrade({
          url: req.getUrl()
          },
          req.getHeader('sec-websocket-key'),
          req.getHeader('sec-websocket-protocol'),
          req.getHeader('sec-websocket-extensions'),
          context);
      },

      message: (ws, message, isBinary) => {
        ws.context.mySuperHeader // <---------------------------
      },
  }
```

I'm sorry if it is a stupid questions, I just cannot find a documented way to do it.

**Comment by streamich** [comment]
I see in C++ version there is `req` param in `open` callback.

![image](https://user-images.githubusercontent.com/9773803/93309117-6ddcd200-f803-11ea-8735-6fcdbff4745e.png)

If something like that was in Node version that would solve it, but I cannot find one.

**Comment by streamich** [comment]
Unfortunately I'm not a C++ developer, but maybe something like this could be a patch?

![image](https://user-images.githubusercontent.com/9773803/93311599-7f73a900-f806-11ea-901a-0bb50dff5703.png)

**Comment by streamich** [comment]
In addition to having `req` in `open`, would still be nice to have a way to pass data from `upgrade` to `open`.

**Comment by streamich** [comment]
nvm, found the right way of doing it: https://github.com/uNetworking/uWebSockets.js/issues/339#issuecomment-651874426

**Comment by ghost** [comment, marked as answer]
There are two examples
